### PR TITLE
FINERACT-2090: Refactor and cleanup Loan entity

### PR DESCRIPTION
**Pull request by kjozsa**
## Description

This PR is part of a larger changeset, covered by the ticket FINERACT-2090, aiming to seriously improve the Loan module's code quality. These changes focus solely on cleaning up the Loan class, which was ~7000 lines long at the develop HEAD originally. 

All changes are either of the following kinds:
- introducing Lombok and removing unnecessary code parts of the Loan entity
- rewriting imperative filter/find methods using the stream APIs, hence improving readability and reducing potential bugs, trading in for a small performance cost
- fixing cosmetic and inconsistent formatting issues

Applying these changes helped to cut over 600 lines of code from this class alone, keeping the same functionality and seriously improving readability of the remaining parts.

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [X ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ X] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ X] Create/update unit or integration tests for verifying the changes made.

- [ X] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ X] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ X] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 1]
```diff
@@ -208,106 +187,123 @@ public class Loan extends AbstractAuditableWithUTCDateTimeCustom {
     private ExternalId externalId;
 
     @ManyToOne
-    @JoinColumn(name = "client_id", nullable = true)
+    @JoinColumn(name = "client_id")
     private Client client;
 
+    @Getter
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: We have Lombok getter setter on the class. if we would like to manually decide on which field we would like to generate one, we should remove that one. Thoughts?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 2]
```diff
@@ -769,10 +773,6 @@ public ChangedTransactionDetail reprocessTransactions() {
      *
      * If "suppliedTransactionDate" is not passed Id, the transaction date is set to the loans due date if the due date
      * is lesser than todays date. If not, the transaction date is set to todays date
-     *
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:773]: Was it incorrect or why did you remove?

Comment (kjozsa) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:773]: because the empty javadocs tags did not add any value imho

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 3]
```diff
@@ -3286,60 +3063,36 @@ private void handleLoanOverpayment(LocalDate transactionDate, final LoanLifecycl
...
-            if (previousTransaction.isRepaymentLikeType() && previousTransaction.isNotReversed()
-                    && DateUtils.isBefore(currentTransactionDate, previousTransaction.getTransactionDate())) {
-                isAfterLatRepayment = false;
-                break;
-            }
-        }
-        return isAfterLatRepayment;
+        return loanTransactions.stream() //
+                .filter(t -> t.isRepaymentLikeType() && t.isNotReversed()) //
+                .noneMatch(t -> DateUtils.isAfter(loanTransaction.getTransactionDate(), t.getTransactionDate()));
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Based on the original, i reckon it should be `isEquals` or `isAfter`...

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 4]
```diff
@@ -3286,60 +3063,36 @@ private void handleLoanOverpayment(LocalDate transactionDate, final LoanLifecycl
...
-                                    && ((loanTransaction.getId() == null && previousTransaction.getId() == null)
-                                            || (loanTransaction.getId() != null && (previousTransaction.getId() == null
-                                                    || loanTransaction.getId().compareTo(previousTransaction.getId()) < 0)))))) {
-                isChronologicallyLatestRepaymentOrWaiver = false;
-                break;
-            }
-        }
-        return isChronologicallyLatestRepaymentOrWaiver;
+        return loanTransactions.stream()
+                .noneMatch(previousTransaction -> !previousTransaction.isDisbursement() && !previousTransaction.isReversed()
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Please double check as i believe the negated version would be different:
from < to =>

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 3 above)
Comment (kjozsa) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: nice catch, thank you

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 5]
```diff
@@ -3292,54 +3069,40 @@ private boolean isChronologicallyLatestRepaymentOrWaiver(final LoanTransaction l
...
-                    && DateUtils.isBefore(currentTransactionDate, previousTransaction.getTransactionDate())) {
-                isAfterLatRepayment = false;
-                break;
-            }
-        }
-        return isAfterLatRepayment;
+    private boolean isAfterLastRepayment(final LoanTransaction loanTransaction, final List<LoanTransaction> loanTransactions) {
+        return loanTransactions.stream() //
+                .filter(t -> t.isRepaymentLikeType() && t.isNotReversed()) //
+                .noneMatch(t -> DateUtils.isEqual(loanTransaction.getTransactionDate(), t.getTransactionDate())
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I think here the `allMatch` would be the proper one

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 6]
```diff
@@ -3292,54 +3069,40 @@ private boolean isChronologicallyLatestRepaymentOrWaiver(final LoanTransaction l
...
-            if (previousTransaction.isNotReversed()
-                    && !DateUtils.isAfter(currentTransactionDate, previousTransaction.getTransactionDate())) {
-                isChronologicallyLatestRepaymentOrWaiver = false;
-                break;
-            }
-        }
-        return isChronologicallyLatestRepaymentOrWaiver;
+        return loanTransactions.stream() //
+                .filter(LoanTransaction::isNotReversed) //
+                .noneMatch(t -> DateUtils.isAfter(loanTransaction.getTransactionDate(), t.getTransactionDate()));
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I think here the `allMatch` would be the proper one

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 7]
```diff
@@ -3663,14 +3407,10 @@ private ChangedTransactionDetail closeDisbursements(final ScheduleGeneratorDTO s
...
     public LoanTransaction getLatestTransaction() {
-        LoanTransaction transaction = null;
-        for (LoanTransaction loanTransaction : this.loanTransactions) {
-            if (!loanTransaction.isReversed() && (transaction == null
-                    || DateUtils.isBefore(transaction.getTransactionDate(), loanTransaction.getTransactionDate()))) {
-                transaction = loanTransaction;
-            }
-        }
-        return transaction;
+        return this.loanTransactions.stream() //
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I dont think this is right:
If we have the following list:
LT1: 2024-01-01, ID:1
LT2: 2024-01-02, ID:2
LT3: 2024-01-02, ID:3

It will return LT2 with ID2, but thats not the last one.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 8]
```diff
@@ -4796,33 +4434,11 @@ public LoanTransaction getLastRepaymentOrDownPaymentTransaction() {
     }
 
     public LocalDate getLastUserTransactionForChargeCalc() {
-        LocalDate lastTransaction = getDisbursementDate();
-        if (this.repaymentScheduleDetail().isInterestRecalculationEnabled()) {
-            lastTransaction = getLastUserTransactionDate();
-        }
-        return lastTransaction;
+        return this.repaymentScheduleDetail().isInterestRecalculationEnabled() ? getLastUserTransactionDate() : getLastRepaymentDate();
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I know you will not like this one, but here the `getDisbursementDate()` need to be used and just leave comment, this needs to be reviewed!

**Review by adamsaghy — APPROVED** [review]
LGTM
